wrapper/sessions.go

```diff
@@ -0,0 +1,106 @@
+package wrapper
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"time"
+
+	"nhooyr.io/websocket"
+	"nhooyr.io/websocket/wsjson"
+)
+
+const (
+	gatewayEncoding           = "json"
+	module                    = "github.com/switchupcb/disgo"
+	maxIdentifyLargeThreshold = 250
+	ErrEventUnmarshal         = "an error occurred while unmarshalling a %v Event:\n%w"
+)
+
+// Session represents a session.
+type Session struct {
+	// Token represents the Bot token the session used to connect.
+	Token string
+
+	// ID represents the session ID of the session.
+	ID string
+
+	// Seq represents the last sequence number received by the client.
+	//
+	// https://discord.com/developers/docs/topics/gateway#heartbeat
+	Seq int
+}
+
+// Connect creates open connection to Discord Gateway.
+func (bot *Client) Connect(s *Session) error {
+	// get the Gateway URL.
+	gateway := GetGateway{}
+	response, err := gateway.Send(bot)
+	if err != nil {
+		return fmt.Errorf("an error occurred getting the Gateway API URL\n%w", err)
+	}
+
+	// TODO: zlib compression
+	endpoint := response.URL + "?v=" + VersionDiscordAPI + "&encoding=" + gatewayEncoding
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	// connect to Discord Gateway Websocket.
+	conn, _, err := websocket.Dial(ctx, endpoint, nil)
+	if err != nil {
+		return fmt.Errorf("an error occurred performing WebSocket handshake\n%w", err)
+	}
+
+	defer conn.Close(websocket.StatusInternalError, "StatusInternalError")
+
+	// unmarshall json file bytes into hello event object
```

Comment:
We want to specify the step we are handling first.
```
// handle the Hello event received upon connecting.
```

The code shows that you unmarshal the JSON file however — if you wish — you can include that it's a JSON response with something like:
```
// handle the incoming Hello event (JSON) upon connecting to the socket.
```
_or Gateway or whatever indicates that it's the WebSocket we are connecting to._